apis/datadoghq/v1alpha1/datadogpodautoscaler_types.go

```diff
@@ -0,0 +1,405 @@
+
+	// Owner defines the source of truth for this object (local or remote)
+	// Value needs to be set when a DatadogPodAutoscaler object is created.
+	Owner DatadogPodAutoscalerOwner `json:"owner"`
+
+	// RemoteVersion is the version of the .Spec currently store in this object.
+	// Only set if the owner is Remote.
+	RemoteVersion *uint64 `json:"remoteVersion,omitempty"`
+
+	// Policy defines how recommendations should be applied.
+	Policy *DatadogPodAutoscalerPolicy `json:"policy,omitempty"`
+
+	// Recommender defines the recommender to use for the autoscaler and its settings.
+	// Default to the `containerMetrics` recommender.
+	Recommender *DatadogPodAutoscalerRecommender `json:"recommender,omitempty"`
+
+	// Constraints defines constraints that should always be respected.
+	Constraints *DatadogPodAutoscalerConstraints `json:"constraints,omitempty"`
+}
+
+// DatadogPodAutoscalerOwner defines the source of truth for this object (local or remote)
+// +kubebuilder:validation:Enum:=Local;Remote
+type DatadogPodAutoscalerOwner string
+
+const (
+	// DatadogPodAutoscalerLocalOwner states that this `DatadogPodAutoscaler` object is created/managed outside of Datadog app.
+	DatadogPodAutoscalerLocalOwner DatadogPodAutoscalerOwner = "Local"
+
+	// DatadogPodAutoscalerLocalOwner states that this `DatadogPodAutoscaler` object is created/managed in Datadog app.
+	DatadogPodAutoscalerRemoteOwner DatadogPodAutoscalerOwner = "Remote"
+)
+
+// DatadogPodAutoscalerPolicy defines how recommendations should be applied.
+type DatadogPodAutoscalerPolicy struct {
+	// DryRun is a global setting to entirely skip actions on the target resource.
+	// It's also possible to selectively deactivate upscale, downscale or update actions thanks to the `Upscale`, `Downscale` and `Update` fields.
+	// Default to false.
+	DryRun bool `json:"dryRun,omitempty"`
+
+	// Update defines the policy to update target resource.
+	Update *DatadogPodAutoscalerUpdatePolicy `json:"update,omitempty"`
+
+	// Upscale defines the policy to scale up the target resource.
+	Upscale *DatadogPodAutoscalerScalingPolicy `json:"upscale,omitempty"`
+
+	// Downscale defines the policy to scale up the target resource.
+	Downscale *DatadogPodAutoscalerScalingPolicy `json:"downscale,omitempty"`
+}
+
+// DatadogPodAutoscalerUpdateMode defines the mode of the update policy.
+// +kubebuilder:validation:Enum:=Auto;DryRun
+type DatadogPodAutoscalerUpdateMode string
+
+const (
+	// DatadogPodAutoscalerAutoMode is the default mode.
+	DatadogPodAutoscalerAutoMode DatadogPodAutoscalerUpdateMode = "Auto"
+
+	// DatadogPodAutoscalerDrynRunMode will skip the update of the target resource.
+	DatadogPodAutoscalerDrynRunMode DatadogPodAutoscalerUpdateMode = "DryRun"
+)
+
+// DatadogPodAutoscalerUpdatePolicy defines the policy to update target resource.
+type DatadogPodAutoscalerUpdatePolicy struct {
+	// Mode defines the mode of the update policy.
+	Mode DatadogPodAutoscalerUpdateMode `json:"mode,omitempty"`
+}
```

Comment:
I think `policy.dryRun: true/false` will be confused with `policy.update.mode: DryRun/Auto`. I think we should align the `update` policy with the `upscale` and `downscale` policies, and make it `policy.update.strategy: disabled/auto`.

Reply:
Implemented